crates/data_components/src/dynamodb/stream.rs

```diff
@@ -0,0 +1,603 @@
+use arrow_array::RecordBatch;
+use arrow_array::builder::{ArrayBuilder, ListBuilder, StringBuilder};
+use aws_sdk_dynamodb::types::AttributeValue as DynamoDbAttributeValue;
+use aws_sdk_dynamodbstreams::types::AttributeValue as StreamsAttributeValue;
+use aws_sdk_dynamodbstreams::types::OperationType;
+use dynamo_subscriber::error::StreamResult;
+use snafu::prelude::*;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+pub fn process_batch(
+    batch: StreamResult,
+    table_schema: &Arc<Schema>,
+    primary_keys: &[String],
+    unnest_depth: Option<usize>,
+) -> Result<ChangeEnvelope, StreamError> {
+    let changes_schema = changes_schema(table_schema).clone();
+
+    let batch = batch.map_err(|e| StreamError::FailedToReceiveMessage { source: e })?;
+
+    let mut changes_struct_builder =
+        StructBuilder::from_fields(changes_schema.fields().clone(), batch.len());
+
+    for record in batch {
+        let (op_str, item_data) = match (&record.event_name, &record.dynamodb) {
+            (Some(event_name), Some(dynamodb)) => match event_name {
+                OperationType::Insert | OperationType::Modify => {
+                    let Some(item) = &dynamodb.new_image else {
+                        continue;
+                    };
+                    let streams_item = streams_to_dynamodb_item(item.clone());
+
+                    let (unnested_streams_item, _) = match unnest_depth {
+                        None => (streams_item, HashSet::new()),
+                        Some(depth) => unnest_dynamodb_item(&streams_item, depth).map_err(|e| {
+                            StreamError::DeserializationError {
+                                message: e.to_string(),
+                            }
+                        })?,
+                    };
+
+                    let op = if matches!(event_name, OperationType::Insert) {
+                        "c"
+                    } else {
+                        "u"
+                    };
+
+                    (op, unnested_streams_item)
+                }
+                OperationType::Remove => {
+                    let Some(keys_item) = &dynamodb.keys else {
+                        continue;
+                    };
+                    let streams_keys_item = streams_to_dynamodb_item(keys_item.clone());
+                    ("d", streams_keys_item)
+                }
+                _ => continue,
+            },
+            _ => continue,
+        };
+
+        // Append row to changes struct
+        changes_struct_builder.append(true);
+
+        // Populate each field in the changes schema
+        for (idx, field) in changes_schema.fields().iter().enumerate() {
+            let field_builder = changes_struct_builder.field_builder_array(idx);
+
+            match field.name().as_str() {
+                "op" => {
+                    let str_builder =
+                        downcast_builder::<StringBuilder>(field_builder).map_err(|e| {
```

Comment:
The `context` method for Snafu should allow you to save a few lines of code when handling the errors on these, i.e.:

```rust
downcast_builder::<StringBuilder>(field_builder).context(DeserializationSnafu)
```

You will need to add the `pub` visibility to the Snafu error to be able to use it across modules (https://docs.rs/snafu/latest/snafu/derive.Snafu.html#controlling-visibility). And you will need to set the `source`  to the specific error type of the downcast.